vertx-web-session-stores/vertx-web-sstore-redis/src/main/java/io/vertx/ext/web/sstore/redis/impl/RedisSessionStoreImpl.java

```diff
@@ -0,0 +1,292 @@
+        processHandler.handle(connResult.result());
+      }
+    });
+  }
+
+  private <R, E> void wrapResult(AsyncResult<R> asResult, E result, Handler<AsyncResult<E>> resultHandler) {
+    if (asResult.failed()) {
+      resultHandler.handle(Future.failedFuture(asResult.cause()));
+    } else {
+      resultHandler.handle(Future.succeededFuture(result));
+    }
+  }
+
+  /**
+   * A helper class to Redis to be able to auto reconnect.
+   */
+  private static class RedisClientHelper {
+    private static final Logger LOG = LoggerFactory.getLogger(RedisClientHelper.class.getName());
+    // Attempt to reconnect up to MAX_RECONNECT_RETRIES
+    private static final int MAX_RECONNECT_RETRIES = 16;
+    private static final String CONNECT_LOCK = "vertx.sstore.redis.connect";
+
+    private final Vertx vertx;
+    private final Redis client;
+    private RedisConnection connection;
+    private long retryTimerId = -1;
+    private final AtomicBoolean isConnected = new AtomicBoolean(false);
+
+    private RedisClientHelper(Vertx vertx, RedisOptions options) {
+      this.vertx = vertx;
+      client = Redis.createClient(vertx, options);
+    }
+
+    void close() {
+      if (retryTimerId != -1) {
+        vertx.cancelTimer(retryTimerId);
+      }
+      client.close();
+      isConnected.set(false);
+    }
+
+    /**
+     * Will create a redis connection if needed and
+     * setup a reconnect handler when there is
+     * an exception related to.
+     */
+    Future<RedisConnection> getConnection() {
+      if (isConnected.get()) {
+        return Future.succeededFuture(connection);
+      }
+
+      Promise<RedisConnection> promise = Promise.promise();
+
+      vertx.sharedData().getLocalLock(CONNECT_LOCK, lockRes -> {
+        if (lockRes.failed()) {
+          promise.fail(lockRes.cause());
+          return;
+        }
+
+        if (isConnected.get()) {
+          promise.complete(connection);
+          lockRes.result().release();
+          return;
+        }
+
+        client.connect(onConnect -> {
+          if (onConnect.succeeded()) {
+            connection = onConnect.result();
+            isConnected.set(true);
+
+            // make sure the client is reconnected on error
+            connection.exceptionHandler(e -> {
+              isConnected.set(false);
+              tryReconnect(0);
+            });
+
+            promise.complete(connection);
+          } else {
+            promise.fail(onConnect.cause());
+          }
+
+          lockRes.result().release();
+        });
+      });
+
+      return promise.future();
+    }
+
+    private void tryReconnect(int retry) {
+      if (retry > MAX_RECONNECT_RETRIES) {
+        LOG.error(String.format(
+          "Redis session server can't be reached after %s retries.",
+          MAX_RECONNECT_RETRIES));
+      } else {
+        // retry with increased delay
+        long delay = (long) (Math.pow(2, MAX_RECONNECT_RETRIES - Math.max(MAX_RECONNECT_RETRIES - retry, 9)) * 10);
```

Comment:
`Math.pow(2, MAX_RECONNECT_RETRIES - Math.max(MAX_RECONNECT_RETRIES - retry, 9)` can be reduced to `Math.pow(2, Math.min(MAX_RECONNECT_RETRIES - 9, retry)`.